contracts/pools/weighted-oracle/WeightedOracleMath.sol

```diff
@@ -14,60 +14,88 @@
+        }
     }
 
-    function _calculateSpotPriceLn(
+    /**
+     * @dev Restores `value` from logarithmic space. `value` is expected to be the result of a call to `_toLowResLog`,
+     * any other function that returns 4 decimals fixed point logarithms, or the sum of such values.
+     */
+    function _fromLowResLog(int256 value) internal pure returns (uint256) {
+        return uint256(LogExpMath.exp(value * _COMPRESSION_FACTOR));
+    }
+
+    /**
+     * @dev Calculates the logarithm of the spot price of token B in token A.
+     *
+     * The return value is a 4 decimal fixed-point number: use `_fromLowResLog` to recover the original value.
+     */
+    function _calcLogSpotPrice(
         uint256 normalizedWeightA,
         uint256 balanceA,
         uint256 normalizedWeightB,
         uint256 balanceB
-    ) internal pure returns (int256 spotPriceLn) {
-        //Rounding direction does not matter because we are compressing the log result at the end.
-        uint256 spotPrice = balanceA.divDown(normalizedWeightA).divDown(balanceB.divUp(normalizedWeightB));
-
-        spotPriceLn = SignedFixedPoint.ln(spotPrice.toInt256()) / _PRICE_COMPRESSION_FACTOR;
+    ) internal pure returns (int256) {
+        // Max balances are 2^112 and min weights are 0.01, so balance / weight can always be computed.
+        // Rounding direction is irrelevant as we're about to introduce much larger error when converting to log space:
+        // we use divDown because it uses less gas.
+        uint256 spotPrice = balanceA.divDown(normalizedWeightA).divDown(balanceB.divDown(normalizedWeightB));
+        return _toLowResLog(spotPrice);
     }
 
-    function _calculateBPTPriceLn(
+    /**
+     * @dev Calculates the price of BPT in a token. `logBptTotalSupply` should be the result of calling `_toLowResLog`
+     * with the current BPT supply.
+     *
+     * The return value is a 4 decimal fixed-point number: use `_fromLowResLog` to recover the original value.
+     */
+    function _calcLogBPTPrice(
         uint256 normalizedWeight,
         uint256 balance,
-        int256 bptTotalSupplyLn
-    ) internal pure returns (int256 bptPriceLn) {
-        //Rounding direction does not matter because we are compressing the log result at the end.
-        int256 totalBptLn = SignedFixedPoint.ln(balance.divDown(normalizedWeight).toInt256());
+        int256 logBptTotalSupply
+    ) internal pure returns (int256) {
+        // BPT price = (balance / weight) / total supply
+        // Since we already have ln(total supply) and want to compute ln(BPT price), we perform the computation in log
+        // space directly: ln(BPT price) = ln(balance / weight) - ln(total supply)
 
-        bptPriceLn = totalBptLn / _INVARIANT_COMPRESSION_FACTOR - bptTotalSupplyLn;
-    }
+        // Rounding direction is irrelevant as we're about to introduce much larger error when converting to log space:
+        // we use divDown because it uses less gas.
```

Comment:
```suggestion
        // The rounding direction is irrelevant as we're about to introduce a much larger error when converting to log
        // space: we use `divDown` because it uses less gas.
```